python_modules/libraries/dagster-aws/dagster_aws/s3/sensor.py

```diff
@@ -1,43 +1,100 @@
-from typing import Optional
+from datetime import datetime
+from typing import Any, List, Optional, cast
 
 import boto3
 import dagster._check as check
+from dagster._annotations import deprecated
+
+from dagster_aws._stubs import ObjectTypeDef
 
 
 class ClientException(Exception):
     pass
 
 
-def get_s3_keys(
+def get_objects(
     bucket: str,
     prefix: str = "",
     since_key: Optional[str] = None,
-    s3_session: Optional[boto3.Session] = None,
-):
+    since_last_modified: Optional[str] = None,
+    client=None,
+) -> List[ObjectTypeDef]:
+    """Retrieves a list of object keys in S3 for a given `bucket`, `prefix`, and filter option.
+
+    Args:
+        bucket (str): s3 bucket
+        prefix (str): s3 object prefix
+        since_key (Optional[str]): retrieve objects after the modified date of this key
+        since_last_modified (Optional[str]): retrieve objects after this timestamp
+        client (Optional[boto3.Client]): s3 client
+
+    Returns:
+        List of object keys in S3.
+
+    """
     check.str_param(bucket, "bucket")
     check.str_param(prefix, "prefix")
     check.opt_str_param(since_key, "since_key")
+    check.opt_str_param(since_last_modified, "since_last_modified")
 
-    if not s3_session:
-        s3_session = boto3.client("s3", use_ssl=True, verify=True)
+    if not client:
+        client = boto3.client("s3")
 
-    if not s3_session:
+    if not client:
         raise ClientException("Failed to initialize s3 client")
 
-    paginator = s3_session.get_paginator("list_objects_v2")  # type: ignore
-    page_iterator = paginator.paginate(Bucket=bucket, Prefix=prefix)
+    paginator = client.get_paginator("list_objects_v2")
+    page_iterator = paginator.paginate(Bucket=bucket, Prefix=prefix, Delimiter="")
 
-    contents = []
+    objects: List[ObjectTypeDef] = []
     for page in page_iterator:
-        contents.extend(page.get("Contents", []))
+        contents = page.get("Contents", [])
+        objects.extend([cast(ObjectTypeDef, obj) for obj in contents])
+
+    if since_key and not any(obj.get("Key") == since_key for obj in objects):
+        raise Exception("Provided `since_key` is not present in list of objects")
+
+    sorted_objects = [obj for obj in sorted(objects, key=lambda x: x.get("LastModified"))]
+
+    if since_key and since_key:
+        for idx, obj in enumerate(sorted_objects):
+            if obj.get("Key") == since_key:
+                return sorted_objects[idx + 1 :]
+
+    if since_last_modified:
+        since_last_modified_dt = datetime.fromisoformat(since_last_modified)
+        for idx, obj in enumerate(sorted_objects):
+            if obj.get("LastModified") >= since_last_modified_dt:
+                return sorted_objects[idx + 1 :]
```

Comment:
Sorry if I'm misunderstanding. I think this will miss an item if the `since_last_modified` does not exactly align with an object, for example if the object being used as the cursor was modified again since last run. (I don't see a `since_last_modified` test, so I'm just going by my own example here:

```py
>>> sorted_objs = [1, 2, 3, 4]
>>> since_last_mod = 2.3
>>> for i, obj in enumerate(sorted_objs):
...     if obj >= since_last_mod:
...         print(f"{obj=} at {i=}: {sorted_objs[i + 1 :]=}")
...         break
... 
obj=3 at i=2: sorted_objs[i + 1 :]=[4]  # should be [3, 4]
```

Reply:
Thanks for taking a look at this Mark.

I've changed the `since_last_modified` parameter to be a `datetime` since that is the type that we actually get back from querying objects.

I've also added a test using the `since_last_modified` parameter here:

https://github.com/dagster-io/dagster/pull/25310/files#diff-854445aee1d01040c738b3eb4c659e75397402f2e855d03d9ac9d5abd24679e4R181

If I create object A, then objects B and C, and use the _LastModified_ timestamp of object A, then the method will return objects B, and C. Is this the expected behavior for you? Should a comment be added to indicate that it is non-inclusive?

Reply:
Thanks Colton!

Using a `datetime` in the interface sounds great.

Yes, I would expect the `since_last_modified` to not be inclusive, so I think the case you describe is as expected.

How does this behave though?

create A at t=1
create B at t=2
-> save LastModified timestamp of B (t=2)
create C at t=3
overwrite B at t=4

I would expect `get_objects(since_last_modified=2)` to return `[C, B]`. But I think it may only return `[B]`, because the `obj >= since_last_modified` will only be true once it reaches `C` (since we modified `B`).

Reply:
I see. You are correct that erroneously does not include C because we're doing the index + 1. 🤔

Reply:
I believe I've fixed it by changing this logic:

https://github.com/dagster-io/dagster/pull/25310/files#diff-00f6b7fbbe8279a609de0de9f35d317b73742ecb02d6cc2399c948ee5754c0aaR66-R67